controllers/scenario/scenario_adapter.go

```diff
@@ -0,0 +1,164 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scenario
+
+import (
+	"context"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/go-logr/logr"
+	applicationapiv1alpha1 "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	"github.com/redhat-appstudio/integration-service/api/v1alpha1"
+	"github.com/redhat-appstudio/integration-service/gitops"
+	"github.com/redhat-appstudio/operator-goodies/reconciler"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Adapter holds the objects needed to reconcile a Release.
+type Adapter struct {
+	application *applicationapiv1alpha1.Application
+	scenario    *v1alpha1.IntegrationTestScenario
+	logger      logr.Logger
+	client      client.Client
+	context     context.Context
+}
+
+// NewAdapter creates and returns an Adapter instance.
+func NewAdapter(application *applicationapiv1alpha1.Application, scenario *v1alpha1.IntegrationTestScenario, logger logr.Logger, client client.Client,
+	context context.Context) *Adapter {
+	return &Adapter{
+		application: application,
+		scenario:    scenario,
+		logger:      logger,
+		client:      client,
+		context:     context,
+	}
+}
+
+// EnsureCreatedScenarioIsValid is an operation that ensures created IntegrationTestScenario is valid
+// in case it is, set its owner reference
+func (a *Adapter) EnsureCreatedScenarioIsValid() (reconciler.OperationResult, error) {
+
+	// First check if application exists or not
+	if a.application == nil {
+		a.logger.Info("Application for scenario was not found.",
+			"Sceario.Name", a.scenario.Name)
+
+		patch := client.MergeFrom(a.scenario.DeepCopy())
+		SetScenarioIntegrationStatusAsInvalid(a.scenario, "Failed to get application for scenario.")
+		err := a.client.Status().Patch(a.context, a.scenario, patch)
+		if err != nil {
+			a.logger.Error(err, "Failed to update Scenario",
+				"Scenario.Name", a.scenario.Name)
+			return reconciler.RequeueWithError(err)
+		}
+		return reconciler.ContinueProcessing()
+	}
+	// Checks if scenario has ownerReference assigned to it
+	if a.scenario.OwnerReferences == nil {
+		patch := client.MergeFrom(a.scenario.DeepCopy())
+		err := ctrl.SetControllerReference(a.application, a.scenario, a.client.Scheme())
+		if err != nil {
+			a.logger.Error(err, "Error setting owner reference.",
+				"Application.Name", a.application.Name,
+				"Application.Namespace", a.application.Namespace)
+			patch := client.MergeFrom(a.scenario.DeepCopy())
+			SetScenarioIntegrationStatusAsInvalid(a.scenario, "Failed to get application for scenario.")
```

Comment:
I don't think we want to set the scenario as invalid because we failed when setting the owner reference, that part's our fault, so logging the issue and calling the `RequeueWithError` should be enough.

Reply:
Agreed.